File: UniquePathsII.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid):
        """
        :type obstacleGrid: List[List[int]]
        :rtype: int
        """
        m = len(obstacleGrid)
        n = len(obstacleGrid[0])
        
        temp = [0 for i in range(m)]
        temp[0] = 1
        for j in range(n):
            for i in range(m):
                if j==0 and i==0:
                    if obstacleGrid[i][j]:
                        temp[i] = 0
                    continue
                if obstacleGrid[i][j]:
                    temp[i] = 0
                else:
                    temp[i] = (temp[i] if j-1>=0 else 0) + (temp[i-1] if i-1>=0 else 0)
        return temp[m-1]
        
if __name__=='__main__':
    m = int(input())
    obstacleGrid = []
    for i in range(m):
        row = list(input())
        obstacleGrid.append(row)
        
    solution = Solution()
    print(solution.uniquePathsWithObstacles(obstacleGrid))

File: UniquePathsII.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int> >& obstacleGrid) {
        int m=obstacleGrid.size();
        int n=obstacleGrid[0].size();
        int temp[m][n];
        temp[0][0]=1;
        
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(i==0&&j==0){
                    if(obstacleGrid[i][j]) temp[i][j]=0;
                    continue;
                }
                if(obstacleGrid[i][j]) temp[i][j]=0;
                else{
                    temp[i][j]=(j-1>=0 ? temp[i][j-1] : 0) + (i-1>=0 ? temp[i-1][j] : 0);
                }
            }
        }
        
        return temp[m-1][n-1];
    }
};

int main(){
    int m, n;
    cin>>m>>n;
    
    vector<vector<int> > obstacleGrid;
    
    for(int i=0;i<m;i++){
        vector<int> row;
        for(int j=0;j<n;j++){
            int grid;
            cin>>grid;
            row.push_back(grid);
        }
        obstacleGrid.push_back(row);
    }
    
    Solution *solution=new Solution();
    cout<<solution->uniquePathsWithObstacles(obstacleGrid);
    
    return 0;
}
